sql/rls: prevent leak of hidden rows in RLS due to predicate reordering

spilchen opened the pull request:
RLS policies are applied as filters to scan operations before any user-defined predicates. Previously, the optimizer could reorder these predicates freely, which could result in information leakage: users could infer the existence of hidden rows based on query behavior.

This change wraps the RLS filter in a Barrier operator, which prevents it from being reordered across non-leak-proof expressions. This ensures that evaluation order is preserved and RLS protections remain intact.

 The Barrier is marked as permeable, allowing optgen rules to push the Barrier up the plan tree for expressions that are leakproof. Only optgen rules for the Select operator were added in this change. Subsequent changes will handle joins and projections.

Informs #146952

Epic: CRDB-48807
Release note (bug fix): Fixed a security issue where optimizer predicate reordering could leak information about hidden rows protected by RLS policies.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/select.go` line 1370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORJDz2uDUhHhhU52MZ8:-ORJDz2uDUhHhhU52MZ9:b-4ctr6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a0778fc7f5e292036eaa7cbf1c1b79a0798808/pkg/sql/opt/optbuilder/select.go#L1370)):*
> ```Go
> 			b.addBarrier(inScope)
> 		}
> 	}
> ```

Drive-by review: I think this is incomplete - something like this would prevent us adding a barrier:
```
SELECT * FROM t t1 INNER JOIN t t2 ON t1.y = 20 AND t2.y/0 = 0;
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/optbuilder/select.go` line 1370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORJDz2uDUhHhhU52MZ8:-ORM9_P1Dv2KwxUs61eX:bnt98yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a0778fc7f5e292036eaa7cbf1c1b79a0798808/pkg/sql/opt/optbuilder/select.go#L1370)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Drive-by review: I think this is incomplete - something like this would prevent us adding a barrier:
```
SELECT * FROM t t1 INNER JOIN t t2 ON t1.y = 20 AND t2.y/0 = 0;
```
</blockquote></details>

Ah, good point. Any predicate that comes in outside of a WHERE clause will likely have an issue. I'm not sure how best to handle this though. Any suggestions? We don't know when we add the RLS filter if a barrier is needed. It depends on if any subsequent filter is not leakproof. It's like we need general logic for adding a filter:
- if the filter is not leakproof
  - if a RLS predicate was added && no barrier was added
    - add a barrier

I'm not sure if there is a good place for this general logic.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/rules/select.opt` line 455 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORSRGGoDzxzGAk54hm2:-ORSRGGp0yphDj4kzczr:b-ucyio3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1335862dcc2eef49840c169b23772ce9d994a7a/pkg/sql/opt/norm/rules/select.opt#L455)):*
> ```
> # all filter expressions are fully leak-proof and the Barrier is marked for
> # skipping.
> [EliminateSelectBarrier, Normalize]
> ```

Wow, really nice job on these rules!

The issue I see with their current form is that removing the barrier entirely means that a leakproof expression further "down" the plan tree could eliminate the barrier, allowing non-leakproof expressions further "up" the tree to be pushed down. So, I think these rules need to "push" operators below the barrier, rather than eliminating the barrier altogether.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/select.opt` line 455 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORSRGGoDzxzGAk54hm2:-ORW9cX-BipokI2B6Iwv:b2yumyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1335862dcc2eef49840c169b23772ce9d994a7a/pkg/sql/opt/norm/rules/select.opt#L455)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Wow, really nice job on these rules!

The issue I see with their current form is that removing the barrier entirely means that a leakproof expression further "down" the plan tree could eliminate the barrier, allowing non-leakproof expressions further "up" the tree to be pushed down. So, I think these rules need to "push" operators below the barrier, rather than eliminating the barrier altogether.
</blockquote></details>

Thanks for that clarification. Makes sense. I adjusted the rules so that it pushes the barrier up rather than eliminating them.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
Looking good - but I only reviewed the select rule and the duplicate barrier rule. IMO it would be good to focus on the basics in the first commit or PR before adding rules for joins and projects - just to keep the review focused.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-:-ORwqCJe4di5YBF7SieR:bjx9hrs)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @spilchen)

___
*[`pkg/sql/opt/norm/rules/barrier.opt` line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwpgzq2eVTB0sLC6AH:-ORwpgzq2eVTB0sLC6AI:bpmxuji) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/barrier.opt#L16)):*
> ```
> )
> =>
> (Barrier $input $outerBarrierPrivate)
> ```

We should have some tests for this rule. If it's too hard to construct a query that has duplicate barriers, we can use `exprnorm`. Here's an example: https://github.com/cockroachdb/cockroach/blob/da630eb83e80e5eae515a06406f53ffcf7028b4b/pkg/sql/opt/norm/testdata/rules/reject_nulls#L406-L460

And here's some docs: https://github.com/cockroachdb/cockroach/blob/e7c1c11b208b5ca798b53a6863eec5c61636128f/pkg/sql/opt/optgen/exprgen/expr_gen.go#L30-L73
___
*[`pkg/sql/opt/norm/testdata/rules/select` line 2547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwoeJ_2iY-WjEj8vaP:-ORwoeJ_2iY-WjEj8vaQ:b-v1t28i) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/testdata/rules/select#L2547)):*
> ```
> SELECT * FROM rls WHERE y = 20;
> ----
> barrier
> ```

I think we should try to eliminate barrier expressions at the root, for simplicity. I'm not sure how to do that off the top of my head, but I can experiment with it. You can leave a TODO here, unless @DrewKimball has an idea.
___
*[`pkg/sql/opt/norm/rules/select.opt` line 455 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwpTBxA9eH_30TAKHL:-ORwpTBxA9eH_30TAKHM:b-lvksve) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/select.opt#L455)):*
> ```
> # the Select when all filter expressions are leak-proof. This ensures that
> # the Barrier is preserved in case it’s needed by expressions higher in the tree.
> [PushSelectBarrierAbove, Normalize]
> ```

nit: I would call this `PushFiltersIntoBarrier` to be consistent with naming of other rules.
___
*[`pkg/sql/opt/norm/rules/select.opt` line 458 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwp1DG5ifPiuw5rhHs:-ORwp1DG5ifPiuw5rhHt:bmp3o0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/select.opt#L458)):*
> ```
> (Select
>     (Barrier $input:* $private:* & (IsBarrierSkippable $private))
>     $filters:* & (HasAllLeakProofFilters $filters)
> ```

This misses an opportunity to push some leak proof filters below the barrier when there is a mix of leakproof
___
*[`pkg/sql/opt/ops/relational.opt` line 1545 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwoOI78Z3Q9ofLFHOt:-ORwoOI8DhiVUStG_ESV:b-up4061) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/ops/relational.opt#L1545)):*
> ```
>     # SkipIfLeakProof is an optional flag that if set, won't apply the barrier if
>     # subsequent filters are leakproof.
>     SkipIfLeakProof bool
> ```

nit: same suggestion as above—what about `LeakProofPermeable`?
___
*[`pkg/sql/opt/norm/general_funcs.go` line 1566 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwo9go7hNZJM-oIZNp:-ORwo9go7hNZJM-oIZNq:b-od5h90) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/general_funcs.go#L1566)):*
> ```Go
> // IsBarrierSkippable returns true if the barrier is marked as one that can be
> // skipped.
> func (c *CustomFuncs) IsBarrierSkippable(priv *memo.BarrierPrivate) bool {
> ```

nit: brainstorming another name that's maybe more accurate than "skippable": `IsBarrierLeakproofPermeable`
___
*[`pkg/sql/opt/norm/general_funcs.go` line 1571 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwpEAS4dbDb6A61-sj:-ORwpEAS4dbDb6A61-sk:bfcvkso) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/general_funcs.go#L1571)):*
> ```Go
> 
> // HasAllLeakProofFilters returns true if every filter given is leak-proof.
> func (c *CustomFuncs) HasAllLeakProofFilters(filters memo.FiltersExpr) bool {
> ```

nit: here and elsewhere—we've standardized on "leakproof" being one word elsewhere, so this should be "Leakproof".


<!-- Sent from Reviewable.io -->

mgartner commented:
>This misses an opportunity to push some leak proof filters below the barrier when there is a mix of leakproof

comment was cut off: meant to end with "a mix of leakproof and non-leakproof filters"

mgartner commented:
*[`pkg/sql/opt/norm/rules/select.opt` line 458 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwp1DG5ifPiuw5rhHs:-ORxMI4H4M2iEcLZmWgB:bjcwr58) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/select.opt#L458)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This misses an opportunity to push some leak proof filters below the barrier when there is a mix of leakproof
</blockquote></details>

Here's what a more optimal rule could look like:

```
[PushLeakproofFiltersIntoPermeableBarrier, Normalize]
(Select
    (Barrier $input:* $private:* & (IsBarrierLeakproofPermeable $private))
    $filters:* &
        (Let
            (
                $leakproofFilters
                $remainingFilters
                $ok
            ):(SplitLeakproofFilters
                $filters
            )
            $ok
        )
)
=>
(Select
    (Barrier
        (Select $input $leakproofFilters)
        $private
    )
    $remainingFilters
)
```

And `SplitLeakproofFilters` would only return `ok=true` if there are leakproof filters.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/barrier.opt` line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwpgzq2eVTB0sLC6AH:-OSKFFORB6c7MpqxxBz_:b-xgk5zl) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/barrier.opt#L16)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We should have some tests for this rule. If it's too hard to construct a query that has duplicate barriers, we can use `exprnorm`. Here's an example: https://github.com/cockroachdb/cockroach/blob/da630eb83e80e5eae515a06406f53ffcf7028b4b/pkg/sql/opt/norm/testdata/rules/reject_nulls#L406-L460

And here's some docs: https://github.com/cockroachdb/cockroach/blob/e7c1c11b208b5ca798b53a6863eec5c61636128f/pkg/sql/opt/optgen/exprgen/expr_gen.go#L30-L73
</blockquote></details>

I had some tests for this already, but it was in the join test file. I added the `exprnorm` example you suggested.
___
*[`pkg/sql/opt/norm/rules/select.opt` line 458 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwp1DG5ifPiuw5rhHs:-OSKowPcB-wybh3boYRi:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/rules/select.opt#L458)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Here's what a more optimal rule could look like:

```
[PushLeakproofFiltersIntoPermeableBarrier, Normalize]
(Select
    (Barrier $input:* $private:* & (IsBarrierLeakproofPermeable $private))
    $filters:* &
        (Let
            (
                $leakproofFilters
                $remainingFilters
                $ok
            ):(SplitLeakproofFilters
                $filters
            )
            $ok
        )
)
=>
(Select
    (Barrier
        (Select $input $leakproofFilters)
        $private
    )
    $remainingFilters
)
```

And `SplitLeakproofFilters` would only return `ok=true` if there are leakproof filters.
</blockquote></details>

Good idea.
___
*[`pkg/sql/opt/norm/testdata/rules/select` line 2547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-ORwoeJ_2iY-WjEj8vaP:-OSKL1xH5HzfQdcowg0Q:b-ivxlvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be36b3cef627b994434175b6b951a983ac95fffd/pkg/sql/opt/norm/testdata/rules/select#L2547)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think we should try to eliminate barrier expressions at the root, for simplicity. I'm not sure how to do that off the top of my head, but I can experiment with it. You can leave a TODO here, unless @DrewKimball has an idea.
</blockquote></details>

I'm not sure how to remove this with optgen rules. I wonder if we need to special case this after all the rules have been applied? I will add a TODO inside `rules/barrier`.


<!-- Sent from Reviewable.io -->

spilchen commented:
@mgartner PTAL. I will add the optgen rules for projections and joins in a separate PR.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work. Left some minor comments.

> Reviewed 4 of 17 files at r2, 4 of 12 files at r3, 14 of 14 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-:-OSQvfSi7Du3yK1en8KA:b6k5qon)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @spilchen)

___
*[`pkg/sql/opt/exec/explain/testdata/row_level_security` line 320 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-OSQv0WVA-_QV9cSHaNb:-OSQv0WVA-_QV9cSHaNc:b-8n1ifl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c2c92b83cde3cbdec20be4b34ae9755d51f847a/pkg/sql/opt/exec/explain/testdata/row_level_security#L320)):*
> ```
>           table: writer@writer_pkey
>           spans: 1+ spans
>           locking strength: for update
> ```

Hmm, this worries me. I'm guessing the barrier at the root of the logical expression tree is causing us to miss the prior optimization here? And maybe above too? Can you add TODOs here that mention that? I can look into it.
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 2127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-OSQtRyHCaISsTcNaPmn:-OSQtRyHCaISsTcNaPmo:bbezzi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d72ba3ce03878da88268ef0d22c7dfb2219a37/pkg/sql/opt/optbuilder/testdata/row_level_security#L2127)):*
> ```
> # Barrier needed. Try to expose with a division by zero error.
> norm
> SELECT * FROM ob WHERE y = 20 AND y/0 = 0;
> ```

nit: IMO a lot of these tests are redundant here except the first two and the JOIN one at the end. We can keep them if you'd like though.
___
*[`pkg/sql/opt/optgen/lang/support/vim/cropt.vim` line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147348#-OSQudBH7JJKLPWi-W-j:-OSQudBH7JJKLPWi-W-k:be0v4nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d72ba3ce03878da88268ef0d22c7dfb2219a37/pkg/sql/opt/optgen/lang/support/vim/cropt.vim#L38)):*
> ```vim
> 
> syn keyword operator Subquery SubqueryPrivate Any Exists Variable Const Null True False Placeholder
> syn keyword operator Tuple Projections ColPrivate Aggregations AggregationsItem Filters FiltersItem Barrier
> ```

Thanks for updating this!


<!-- Sent from Reviewable.io -->

mgartner commented:
I made some minor refactors in #148130 as a follow-up to this PR.

spilchen committed:
sql/rls: prevent leak of hidden rows in RLS due to predicate reordering

RLS policies are applied as filters to scan operations before any
user-defined predicates. Previously, the optimizer could reorder these
predicates freely, which could result in information leakage: users
could infer the existence of hidden rows based on query behavior.

This change wraps the RLS filter in a Barrier operator, which prevents
it from being reordered across non-leak-proof expressions. This ensures
that evaluation order is preserved and RLS protections remain intact.

The Barrier is marked as permeable, allowing optgen rules to push the
Barrier up the plan tree for expressions that are leakproof. Only optgen
rules for the Select operator were added in this change. Subsequent
changes will handle joins and projections.

Informs #146952

Epic: CRDB-48807
Release note (bug fix): Fixed a security issue where optimizer
predicate reordering could leak information about hidden rows protected
by RLS policies.
